packages/data-context/src/actions/DataEmitterActions.ts

```diff
@@ -62,34 +71,60 @@ export class DataEmitterActions extends DataEmitterEvents {
    * when subscribing, we want to execute the operation to get the up-to-date initial
    * value, and then we keep a deferred object, resolved when the given emitter is fired
    */
-  subscribeTo (evt: keyof DataEmitterEvents, sendInitial = true): AsyncIterator<any> {
+  subscribeTo (evt: keyof DataEmitterEvents, sendInitial = true): AsyncGenerator<any> {
     let hasSentInitial = false
     let dfd: pDefer.DeferredPromise<any> | undefined
-
-    function subscribed (val: any) {
-      dfd?.resolve(val)
+    let pending: any[] = []
+    let done = false
+
+    function subscribed (value: any) {
+      // We can get events here before next() is called setting up the deferred promise
+      // If that happens, we will queue them up to be handled once next eventually is called
+      if (dfd) {
+        dfd.resolve({ done: false, value })
+        dfd = undefined
+      } else {
+        pending.push({ done: false, value })
+      }
     }
     this.pub.on(evt, subscribed)
 
     const iterator = {
       async next () {
+        if (done) {
+          return { done: true, value: undefined }
+        }
+
         if (!hasSentInitial && sendInitial) {
           hasSentInitial = true
 
           return { done: false, value: {} }
         }
 
-        dfd = pDefer()
+        if (pending.length === 0) {
+          dfd = pDefer()
 
-        return { done: false, value: await dfd.promise }
+          return await dfd.promise
+        }
+
+        return pending.shift()
+      },
+      throw: async (error: Error) => {
+        throw error
```

Comment:
Does adding `throw` here make a difference? Or was this more useful while debugging the timing issue with the deferreds?

Reply:
This was to complete the expected type for the async generator. From what I could tell in the graphql-ws an async generator seemed to be what they were expecting if you were going to expose `return` like we are:

https://github.com/enisdenjo/graphql-ws/blob/7649ad739291a6bfb031bdc4b3487291568fe8a8/src/server.ts#L827-L828

Declaring it as an async generator like I am now, allows me to iterate like I am in the test i wrote but it also necessitates adding the throw function.